src/topology/connected.lean

```diff
@@ -132,6 +133,73 @@ begin
     Hs.is_preconnected Ht.is_preconnected
 end
 
+/-- The filtered sUnion of a set S of preconnected subsets is preconnected.
+(A set of sets is filtered if if the inclusion relation is total on it.)
+-/
+theorem is_preconnected.filtered_sUnion_of {S : set (set α)}
+  (K : ∀ s t, s ∈ S → t ∈ S → s ⊆ t ∨ t ⊆ s)
+  (H : ∀ s ∈ S, is_preconnected s ) : is_preconnected ⋃₀ S :=
+begin
+  rw is_preconnected,
+  rintros u v hu hv Huv ⟨a, ⟨s, hsS, has⟩, hau⟩ ⟨b, ⟨t, htS, hbt⟩, hbv⟩,
+  revert a b u v,
+  /- WLOG, we will only prove the case where s ⊆ t -/
+  wlog Hst: s ⊆ t := K s t hsS htS using [s t, t s] tactic.skip,
+  intros a b u v hu hv Huv hau has hbv hbt,
+  have hat : a ∈ t := Hst has,
+  have Hnuv : (t ∩ (u ∩ v)).nonempty,
+    from H t htS u v hu hv ((subset_sUnion_of_mem htS).trans Huv) ⟨a, hat, hau⟩ ⟨b, hbt, hbv⟩,
+  have Ktuv : ( t ∩ (u ∩ v)) ⊆ sUnion S ∩ (u ∩ v),
+    from inter_subset_inter_left _ (subset_sUnion_of_mem htS),
+  exact Hnuv.mono Ktuv,
+
+  /- We now prove t ⊆ s by symmetry -/
+  intros s t a b u v hu hv Huv hau has hbv hbt,
+  rw union_comm at Huv,
+  rw inter_comm u v,
+  exact this t s b a v u hv hu Huv hbv hbt hau has,
+end
+
+/--
+The union of an increasing family s of preconnected sets is preconnected.
+-/
+variables {ι : Type*} [linear_order ι]
+theorem is_preconnected.Union_of_monotone {s : ι → set α} (K : monotone s)
+  (H : ∀ n: ι, is_preconnected (s n)) :
+  is_preconnected ⋃ i, s i :=
+begin
+  rw ← sUnion_range s,
+  apply is_preconnected.filtered_sUnion_of,
+  { rintros - - ⟨i, rfl⟩ ⟨j, rfl⟩,
+    exact (le_total i j).elim (λ h, or.inl $ K h) (λ h, or.inr $ K h) },
+  { rintros - ⟨i, rfl⟩,
+    exact H i }
+end
+
+/--
+If `s` is a family of preconnected sets indexed by `ℕ` such that `s j` meets `s (j+1)`
+for all `j`, then `⋃ n, s n` s is preconnected. -/
+theorem is_preconnected.union_of_chain { s : ℕ → set α }
+  (H : ∀ n: ℕ, is_preconnected (s n))
+  (K : ∀ n: ℕ, (s n ∩ s(n.succ)).nonempty) :
+  is_preconnected (⋃ n, s n) :=
```

Comment:
I feel like this lemma should be generalized. With lemmas like this, it's really tricky to apply this lemma if you want a different indexing type, like Patrick's proof of the finite case shows. Therefore, you want to try to generalize a lemma like this to arbitrary indexing types.

It is a little tricky to state this lemma for an arbitrary indexing type, but I think the following is pretty nice:
```lean
theorem is_preconnected.Union_of_chain' {ι} {s : ι → set α}
  (H : ∀ i, is_preconnected (s i))
  (K : ∀ i j, relation.trans_gen (λ i j : ι, (s i ∩ s j).nonempty) i j) :
  is_preconnected (⋃ n, s n)
```
In words, condition `K` states: the transitive closure of the relation `(λ i j : ι, (s i ∩ s j).nonempty)` is the full relation.

Together with a lemma like the following (which I think doesn't exist yet):
```lean
lemma relation.trans_gen_nat (r : ℕ → ℕ → Prop) {n m : ℕ} 
  (h : ∀ i ∈ Ico (min n m) (max n m), r i i.succ ∧ r i.succ i) : relation.trans_gen r n m := 
sorry
```
we can easily prove your original lemma
```lean
theorem is_preconnected.union_of_chain { s : ℕ → set α }
  (H : ∀ n: ℕ, is_preconnected (s n))
  (K : ∀ n: ℕ, (s n ∩ s(n.succ)).nonempty) :
  is_preconnected (⋃ n, s n) :=
is_preconnected.Union_of_chain' H $ λ i j, 
  relation.trans_gen_nat _ (λ i _, ⟨K i, by { rw inter_comm, exact K i }⟩)
```
I'm sure that with this formulation, something like Patrick's case is also easy to prove.

I'm happy to help with proving this more general case, if you want.

Reply:
To prove this more general version, we probably need a Zorn-like argument.

Reply:
@fpvandoorn (I had answered your remark by email, but it is better to write it here.)

I like your formulation in so that it exactly says that the union a family of preconnected sets indexed by the vertices of a connected graph is preconnected if the sets indexed by any two adjacent vertices meet.

Together with a list of relevant graphs, this would be fine.

By the way, the case (in mathlib) of a union of connected sets with a given common point is a particular case of this (the graph contains a star), and the other interesting graphs to consider are intervals of Z.

Reply:
Also, it might have been easier to prove this theorem using the criterion that any continuous map to the {0,1} -space is constant.

Reply:
@fpvandoorn, I could prove the theorem along the lines you indicate, but I have made a detour through `refl_trans_gen`, whose induction principle was clearly stated in mathlib ('logic/relation.lean'), and this forced me to treat the case of connected sets.
The difference is of very little importance, but one never knows.
(One needs to assume that \iota is nonempty.)

(I still need to add the lemma that the transitive closure of the succ relation on any interval of Z is the full relation.)

Reply:
@fpvandoorn, I could prove the theorem along the lines you indicate, but I have made a detour through `refl_trans_gen`, whose induction principle was clearly stated in mathlib ('logic/relation.lean'), and this forced me to treat the case of connected sets.
The difference is of very little importance, but one never knows.
(One needs to assume that \iota is nonempty.)

(I still need to add the lemma that the transitive closure of the succ relation on any interval of Z is the full relation.)

Reply:
Good! Can you push your work to this branch so that we can review it?

It's probably slightly nicer to do it with `trans_gen` instead of `refl_trans_gen`, so that you don't have to assume non-emptyness. You can do induction on either predicate by doing `induction h` if `h` has type `trans_gen x y` (or `refl_trans_gen x y`). The two induction principles `refl_trans_gen.head_induction_on` and `refl_trans_gen.trans_induction_on` are alternative induction principles (we could easily prove similar results for `trans_gen` if you need them). 
But this is minor, and doesn't necessary have to change for this PR.

Reply:
I can try to do this, but I don't know how to do it. Can I  just add a  new bunch of functions below the previous ones, with new names, and commit/push the whole stuff?

Reply:
Yes, that sounds good. I'll clean it up a bit!

Reply:
The following lemma 
```

lemma succ_gen_of_int_interval (ι:set ℤ) [hι: ord_connected ι]: 
  ∀(i j:ℤ), i ∈ ι → j ∈ ι → relation.refl_trans_gen
     (λ i j:ℤ, i ∈ ι ∧ j ∈ ι ∧ (j = i+1 ∨ i = j+ 1)) i j
  :=
begin
  intros m n hm, 
  refine int.induction_on' n m _ _ _ ,
  -- case: m m
  { intro _, exact relation.refl_trans_gen.refl, }, 
  -- forward induction : m k → m (k+1)
  { intros k hmk IHk hk',
    apply (relation.refl_trans_gen.cases_tail_iff _ m (k+1)).2,
    have hk : k ∈ ι,
      { apply set.ord_connected.out hι hm hk',
        simp only [zero_le_one, and_true, le_add_iff_nonneg_right, mem_Icc, hmk]},
    apply or.intro_right _ , use k, split,
    exact IHk hk,
    apply and.intro hk, apply and.intro hk', 
    simp only [true_or,eq_self_iff_true] }, 
  -- backward inductin : m k → m (k-1)
  { intros k hmk IHk hk',
    have hk : k ∈ ι,
      { apply set.ord_connected.out hι  hk' hm,
      simp only [hmk, zero_le_one, sub_le_self_iff, and_self, mem_Icc]},
    apply (relation.refl_trans_gen.cases_tail_iff _ m (k-1)).2,
    apply or.intro_right _ , use k, split,
    exact IHk hk,
    apply and.intro hk, apply and.intro hk', 
    simp only [sub_add_cancel, eq_self_iff_true, or_true]}  
end
```
proves that int intervals will satisfy the required property.
However, I can't use it because I cannot write something as :
```
theorem is_connected.Union_nat_of_int_interval 
  { ι : set ℤ} {hι₀:nonempty ι} {hι: set.ord_connected ι} 
  { s : ι → set α }
  (H : ∀ i: ι, is_connected (s i))
  (K : ∀ i: ℤ, i ∈ ι → i.succ ∈ ι → (s i ∩ s i.succ).nonempty) :
  is_connected (⋃ i, s i) :=
```

there is a type mismath at `s i` on line 5. A problem of coercion I don't know how to code in Lean.

Reply:
I modified the proofs to work for `preconnected`. 
I reformulated the `Union` lemma a little bit so that the graph is defined by the non-emptyness conditions.
I also reformulated (generalized) the lemmas about `nat`-relations, and move them to a separate file (a new one, since there was no file that already existed that imported the right files where it was appropriate).
I proved similar lemmas for `int`-relations. The formulation I added should be general enough to reason about any `trans_gen` on integers.

To formulate the right lemma on the integers, you probably want to make `s : \Z → set α` but work with `⋃ i \in \iota, s i`.
The most convenient way to do this is to also prove `is_preconnected.Union_of_trans_gen` for a union of the form `⋃ i \in \iota, s i` (we call these forms of unions `bUnion` for bounded union in the library).

Reply:
I added the corresponding induction principles for `trans_gen` in #10331, but didn't end up using them for this PR.

Reply:
Do you want me to go on or will you do it yourself?

Reply:
If you're happy to try it, please do! `is_preconnected.bUnion_of_trans_gen` should come before `is_preconnected.Union_of_trans_gen` with appropriate extra conditions (everything is restricted to a subset of `\iota`), and then `is_preconnected.Union_of_trans_gen` should be an easy consequence.

Reply:
Good. I'll probably do it tomorrow. Thanks for the help !